### PR TITLE
Adjust shadow intensity based on cloud density.

### PR DESCRIPTION
**Pull request by lhofhansl**
Now that https://github.com/minetest/minetest/pull/11944 is merged, we can use that in minetest_game to set the shadow intensity automatically based on the cloud density.

Shadow intensity becomes as much a game emergence feature as clouds.

### CODE REVIEW

File: mods/weather/init.lua
Comment (appgurueu) [inline mods/weather/init.lua:111]: Why not make this `0.7 * (1 - density)`?

**Review by appgurueu — APPROVED** [review]
Looks good; the luacheck warnings are unrelated.

File: mods/weather/init.lua
Comment (lhofhansl) [reply mods/weather/init.lua:111]: +1 that's better.

**Comment by sfan5** [comment]
This has the same bug as https://github.com/minetest/minetest_game/pull/2936#discussion_r830377297 and cannot be merged in this state.

**Comment by appgurueu** [comment]
> This has the same bug as [#2936 (comment)](https://github.com/minetest/minetest_game/pull/2936#discussion_r830377297) and cannot be merged in this state.

Why? It's reasonable for shadow intensity to be tied to cloud density (perhaps this should even be implemented at an engine level, but giving modders full control is also fine). I assume said `if` just exists because the assumptions used here to infer a reasonable, biome-based cloud density from heat and humidity just don't hold for v6 so this is better left disabled for v6. Why can't the feature be selectively applied to v6 only, like the rest of the weather mod?

**Comment by lhofhansl** [comment]
@sfan5 Correct, it would not work with MapGen v6, but for other MapGens it does work, and people get shadows when they enabled them, and then they even make sense w.r.t. clouds. I'll take over nothing any day.

Alternatively could put in a default for MGv6 or derive the cloud density in some other way.

**Comment by sfan5** [comment]
It is reasonable to tie cloud density to shadow density, it is *not* reasonable to have shadows just not work on singlenode or v6.
Also I'm not sure if tying this to `enable_weather` is a good idea. In either case you'll have to update [settingtypes.txt](https://github.com/minetest/minetest_game/blob/master/settingtypes.txt#L75) too.

**Comment by appgurueu** [comment]
> It is reasonable to tie cloud density to shadow density, it is _not_ reasonable to have shadows just not work on singlenode or v6.

Why won't shadows work there? They should just have the default shadow intensity as it isn't explicitly turned off.

**Comment by sfan5** [comment]
Also: 

> and then they even make sense w.r.t. clouds.

with the weather mod disabled there'd be a random (let's say 50%), why wouldn't there be shadows logically?

> I'll take over nothing any day.

This is not wrong however we can just do it properly from the start.

**Comment by sfan5** [comment]
> They should just have the default shadow intensity as it isn't explicitly turned off.

In the respective PR I have been informed by @x2048 that the shadow API is opt-in. If it isn't then we have a problem.

**Comment by lhofhansl** [comment]
> In the respective PR I have been informed by @x2048 that the shadow API is opt-in. If it isn't then we have a problem.

If the client has shadows enabled, this will indeed turn shadows on. Setting the intensity **is** the API to opt into shadows from the game side (again only if also enabled on the client). MT Game is a game, and yes it should opt into shadows. So, yes, MT-Game will opt into shadows with this.

I'm happy to make it work with MGv6, if we agree that that's right direction.

(If we're new discussing whether MT-Game can turn on shadows then ... thanks but no thanks, have that discussion without me please. In any case this will go into my ever-growing personal patchset/fork.)

**Comment by sfan5** [comment]
> I'm happy to make it work with MGv6, if we agree that that's right direction.

Sure, here's my proposal:
* Add fallback code that sets shadow intensity to some okay default value (0.5)
* Have this code apply for unsupported mapgens (`v6`, `singlenode`)
* Also have this code apply if the user disables the weather mod via setting

**Comment by lhofhansl** [comment]
How about this? No shadow intensity set when weather is disabled. Defaults to 0.5 for mgv6 and singlenode. (It is always required that the client enabled shadows of course)

**Comment by x2048** [comment]
0.5 looks too dark. The original shadow intensity before the API was added was 0.2, which may feel a bit low-contrast. Something between 0.3-0.35 could be an optimal default value, tunable by gamma on the client side.

**Comment by lhofhansl** [comment]
As you all know I liked it when it was a client-side setting. :)  Now games have to think about how it will look on the client.

0.3 is good. How about I set it to 1/3 (0.33)?

**Comment by appgurueu** [comment]
> As you all know I liked it when it was a client-side setting. :) Now games have to think about how it will look on the client.
> 
> 0.3 is good. How about I set it to 1/3 (0.33)?
> 
> ... done

How about another MTG server-side setting ;)

**Comment by lhofhansl** [comment]
> How about another MTG server-side setting ;)

No more settings if we can at all avoid it. :)
The client still has the shadow gamma to adjust.

**Comment by lhofhansl** [comment]
Ping.

File: mods/weather/init.lua
Comment (x2048) [inline mods/weather/init.lua:13]: it is simpler to just write player:set_lighting({ shadows = { intensity = 0.33 } }). The API call keeps values you don't pass.

Comment (lhofhansl) [reply mods/weather/init.lua:13]: Didn't check that... Yeah, that's much cleaner.

**Review by appgurueu — APPROVED** [review]

**Comment by lhofhansl** [comment]
I do not have merge rights on MTG (only on the Minetest core repos it seems), so we need a second approval and then someone to merge.

File: mods/weather/init.lua
Comment (sfan5) [inline mods/weather/init.lua:4]: Needs documentation in `minetest.conf.example` that enable_weather also controls shadows.

File: minetest.conf.example
Comment (sfan5) [inline minetest.conf.example:75]: Well technically it doesn't enable shadow *variation* it enables shadows *at all*

**Review by sfan5 — APPROVED** [review]
LGTM

File: minetest.conf.example
Comment (lhofhansl) [reply minetest.conf.example:75]: Hmm... Fair. I'm not good at wording this.

How about "Enable cloud variation by the 'weather' mod and enable and adjust shadows accordingly" ?

Comment (sfan5) [reply minetest.conf.example:75]: +1 for the latter

**Comment by lhofhansl** [comment]
Wanna amend the last commit with the changed settings text?

**Comment by appgurueu** [comment]
> Wanna amend the last commit with the changed settings text?

Just make a trivial follow-up PR which can be merged instantly. There's no need to rewrite history for this, is there?

**Pull request merged**
